internal/buffer/out_message.go

```diff
@@ -100,51 +74,63 @@ func (m *OutMessage) ShrinkTo(n int) {
 			n,
 			m.Len()))
 	}
-
-	m.payloadOffset = n - OutMessageHeaderSize
+	if n == OutMessageHeaderSize {
+		m.Sglist = nil
+	} else {
+		i := 1
+		n -= OutMessageHeaderSize
+		for len(m.Sglist) > i && n >= len(m.Sglist[i]) {
+			n -= len(m.Sglist[i])
+			i++
+		}
+		if n > 0 {
+			m.Sglist[i] = m.Sglist[i][0 : n]
+			i++
+		}
+		m.Sglist = m.Sglist[0 : i]
+	}
 }
 
 // Append is equivalent to growing by len(src), then copying src over the new
 // segment. Int panics if there is not enough room available.
-func (m *OutMessage) Append(src []byte) {
-	p := m.GrowNoZero(len(src))
-	if p == nil {
-		panic(fmt.Sprintf("Can't grow %d bytes", len(src)))
+func (m *OutMessage) Append(src ...[]byte) {
+	if m.Sglist == nil {
+		m.Sglist = append(m.Sglist, nil)
+		*(*reflect.SliceHeader)(unsafe.Pointer(&m.Sglist[0])) = reflect.SliceHeader{
+			Data: uintptr(unsafe.Pointer(&m.header)),
+			Len:  OutMessageHeaderSize,
+			Cap:  OutMessageHeaderSize,
+		}
 	}
-
-	sh := (*reflect.SliceHeader)(unsafe.Pointer(&src))
-	jacobsa_fuse_memmove(p, unsafe.Pointer(sh.Data), uintptr(sh.Len))
-
+	m.Sglist = append(m.Sglist, src...)
 	return
 }
 
 // AppendString is like Append, but accepts string input.
 func (m *OutMessage) AppendString(src string) {
-	p := m.GrowNoZero(len(src))
-	if p == nil {
-		panic(fmt.Sprintf("Can't grow %d bytes", len(src)))
-	}
-
-	sh := (*reflect.StringHeader)(unsafe.Pointer(&src))
-	jacobsa_fuse_memmove(p, unsafe.Pointer(sh.Data), uintptr(sh.Len))
-
+	m.Append([]byte(src))
 	return
 }
 
 // Len returns the current size of the message, including the leading header.
 func (m *OutMessage) Len() int {
-	return OutMessageHeaderSize + m.payloadOffset
+	if m.Sglist == nil {
+		return OutMessageHeaderSize
+	}
+	r := 0
```

Comment:
Can you add a comment here about `m.Sglist[0]` containing the header (enforced by `Append()`), so that readers don’t need to wonder why `OutMessageHeaderSize` isn’t counted here?

Reply:
Ok, added comments